### ISSUE TITLE
Dependabot removes the indirect Go dependencies that are in use since Dec 2022

### ISSUE THREAD

**Issue by peggy-sun-fp** [dependabot/dependabot-core#6479]
### Package ecosystem

go.mod

### Language version

Go 1.18

### dependabot.yml content

<img width="460" alt="Screen Shot 2023-01-19 at 4 09 27 PM" src="https://user-images.githubusercontent.com/106724673/213388372-c98b5aa9-ae7d-468c-a1ea-12c06c44ba72.png">

### Updated dependency

Every dependency upgrade since Dec 2022, such as upgrading golang.org/x/tools from 0.4.0 to 0.5.0

### What you expected to see, versus what you actually saw

Expected: It should only upgrade `golang.org/x/tools` and its related indirect dependencies.
<img width="1102" alt="Screen Shot 2023-01-19 at 4 11 01 PM" src="https://user-images.githubusercontent.com/106724673/213388644-42bfb989-6cea-40d1-843c-43a7530365ad.png">

Actual: Dependabot upgrades `golang.org/x/tools` and removes a lot of indirect dependencies that are in use, causing CI checks to fail!

<img width="890" alt="Screen Shot 2023-01-19 at 4 11 57 PM" src="https://user-images.githubusercontent.com/106724673/213388829-aa375f76-be68-4a9d-8418-cf357783223d.png">
<img width="679" alt="Screen Shot 2023-01-19 at 4 12 42 PM" src="https://user-images.githubusercontent.com/106724673/213388960-7c4bdab7-32f3-43e8-ad91-97e5bb370021.png">
<img width="900" alt="Screen Shot 2023-01-19 at 4 13 22 PM" src="https://user-images.githubusercontent.com/106724673/213389083-4aa2fd04-4e37-4b32-bcee-3f0655c7de62.png">

Dependabot is broken in our repository, so we need to type `go mod tidy` to restore the dependencies deleted by dependabot.

### Native package manager behavior

When I manually upgraded with the same dependency with the command `go get golang.org/x/tools`, none of the above dependencies were removed. Besides, they were still existing after I typed `go mod tidy`.

**Comment by deivid-rodriguez** [comment]
Hei!

Can you try the [dry-run.rb script](https://github.com/dependabot/dependabot-core/blob/fe24ab9e9914ed2e4ed415b2d1cd4828fee07791/README.md#dry-run-script) against your repo to see if the problem can be reproduced using that?

Alternative, can you provide manifest files so that we can try reproducing it ourselves?

Thanks!

**Comment by jeffwidman** [comment]
This may be caused by a bug in https://github.com/dependabot/dependabot-core/pull/6361... as @deivid-rodriguez said dry-run script is going to be the fastest way to get to the root of the problem, otherwise please provide a public repo that reproduces the problem.

**Comment by peggy-sun-fp** [comment]
Hi @deivid-rodriguez @jeffwidman 
As our repo is private, is it okay that I put `dependabot.yml`, `go.mod`, and `go.sum` in a new public repo for you to test?

**Comment by deivid-rodriguez** [comment]
Absolutely!

**Comment by jeffwidman** [comment]
Yes.

Since the updater runs `go mod tidy`, a common solution is to include a `fake.go` file that "dummy imports" the direct deps so that they wont' get pruned by `tidy`:
```
package fakeimports

import (
	_ "github.com/path/to/package/that/we/don't/want/tidy/to/prune"
	_ "github.com/another/package"
)
```

**Comment by peggy-sun-fp** [comment]
Hi @deivid-rodriguez @jeffwidman ,
I have created a [public repository](https://github.com/peggy-sun-fp/test-dependabot) and executed your dry-run script on both repositories. However, I got a lot of error messages `handled error whilst updating <dependency-url>: git_dependencies_not_reachable {:"dependency-urls"=>["github.com/deliveryhero/*"]}`. I guess it is because the latter dependency is private but not sure how to solve it. The results of dry-run script are also attached in the public repo [here](https://github.com/peggy-sun-fp/test-dependabot/tree/main/dry_run_results), thanks!

**Comment by deivid-rodriguez** [comment]
Thanks for providing that!

Looking at logs, this stands out to me:

```
=== github.com/aws/aws-sdk-go (1.44.171)
 => checking for updates 3/38
 => latest available version is 1.44.189
 => latest allowed version is 1.44.189
 => requirements to unlock: own
 => requirements update strategy: 
 => updating github.com/aws/aws-sdk-go from 1.44.171 to 1.44.189
 => handled error whilst updating github.com/aws/aws-sdk-go: git_dependencies_not_reachable {:"dependency-urls"=>["github.com/deliveryhero/pd-fluid-go-sdk"]}
```

Indeed it would feel like an issue due to a dependency being private, but then, why was dependabot able to figure out it needs to upgrade it, and to which version? I know nothing about to go ecosystem, so don't have the answers, but something feels off.

**Comment by jeffwidman** [comment]
@peggy-sun-fp IIRC, you can workaround the `git_dependencies_not_reachable` issue by setting `LOCAL_GITHUB_ACCESS_TOKEN` to an access token that has access to all those repos:

https://github.com/dependabot/dependabot-core/blob/d6883339398279a0b60913cb50cfda8011f07653/bin/dry-run.rb#L14-L15

As @deivid-rodriguez points out, the dry run output looks suspect, but there's also a chance it's a bug in how dry-run parses the output, not the underlying lib... so it'd be good to get a run that isn't blocked on a permissions issue and see what error it starts hitting.

IF you're still coming up empty after that, then we may need to move this one to a support request since it does sorta smell like a legit bug here that will be easier to debug if we have access to the underlying repo and I assume this is your company's paid account so obviously can't open source it.

But first try it locally with a token that has access to all those repos and see what starts breaking.

**Comment by peggy-sun-fp** [comment]
Hi @deivid-rodriguez @jeffwidman 
Thanks for the tip! The dry-run script worked successfully after adding `LOCAL_GITHUB_ACCESS_TOKEN`. However, the results differ when creating PRs. Dry-run does not remove other dependencies while a new PR does. Upon checking the repository, I found a warning saying `Dependabot ran out of disk space during its update`.

<img width="1018" alt="Screen Shot 2023-02-06 at 2 58 46 PM" src="https://user-images.githubusercontent.com/106724673/216904107-bade4982-ce80-4fec-90c1-f4e8d830b829.png">

**Comment by jeffwidman** [comment]
Hmm... For Dry run, did you also:
* set `GOPRIVATE=*` to bypass the go proxy
* clear the go module cache in between runs?

Sorry I forgot to mention these ahead of time, but they can influence how bugs appear due to caching/proxy stuff... see https://github.com/dependabot/dependabot-core/issues/4536#issuecomment-1004412303 for example. And sometimes other errors start getting exposed.

The "ran out of disk space" is likely a semi-spurious error, unless it repro's consistently I'd ignore it.

**Comment by rujutaghanekar** [comment]
Is there any update on the above issue? We are also facing a similar scenario where `Dependabot` removes an indirect dependency but manually running `go mod tidy` does not.

**Comment by jakecoffman** [comment]
I made a couple of large changes this year that may have fixed it:
- https://github.com/dependabot/dependabot-core/pull/12228
- https://github.com/dependabot/dependabot-core/pull/12230

I saw there was a clean public example linked above: 
- https://github.com/XTLS/Xray-core/pull/3680

I was unable to reproduce the problem using Dependabot CLI and also in a PR:
- https://github.com/dsp-testing/go-unrelated-bump/pull/9/changes

If you come across this issue and think it's still happening please link to a public PR so we can debug the issue.
